Testing-the-engine.md

```diff
@@ -0,0 +1,302 @@
+
+If supporting logic are needed for the test case, they can be added to the
+Android app under-test in https://github.com/flutter/engine/tree/master/testing/scenario_app/android/app/src/main/java/dev/flutter/scenarios
+or to the fake Flutter framework under-test in https://github.com/flutter/engine/tree/master/testing/scenario_app/lib.
+
+As best practice, favor adding unit tests if possible since instrumented tests
+are, by nature, non-hermetic, slow and flaky.
+
+End-to-end tests on Android are currently not running on CI [#55326](https://github.com/flutter/flutter/issues/55326)
+
+## Objective-C - iOS embedding
+
+If you're editing `.h` or `.mm` files in the https://github.com/flutter/engine/tree/master/shell/platform/darwin/ios
+directory, you're working on the iOS embedding which connects the core C++
+engine to the iOS SDK APIs and runtime.
+
+### XCTest unit tests
+
+For testing logic within a class at a unit level, create or add to a XCTestCase
+in the XCTest project.
+
+The iOS unit testing infrastructure is split in 2 different locations. The
+`...Test.mm` files in https://github.com/flutter/engine/tree/master/shell/platform/darwin/ios
+contain the unit tests themselves. The
+https://github.com/flutter/engine/tree/master/testing/ios/IosUnitTests directory
+contains an Xcode container project to execute the test.
+
+See https://github.com/flutter/engine/blob/master/testing/ios/IosUnitTests/README.md
+for details on adding new test files.
+
+The engine repo has a unified build system to build C, C++, Objective-C,
+Objective-C++, and Java files using [GN](https://gn.googlesource.com/gn/) and
+[Ninja](https://ninja-build.org/). Since GN and Ninja has to build the C++
+dependencies that the Objective-C classes reference, the tests aren't built by
+the Xcode project in https://github.com/flutter/engine/tree/master/testing/ios/IosUnitTests.
+
+Instead, the engine provides the script:
+
+    `testing/ios/run_tests.py --type=objc`
+
+to easily build and run the XCTests.
+
+Behind the scenes, it invokes GN and Ninja to build the tests and dependencies
+into a single `.dylib`. Then it uses Xcode and the Xcode project at
+`testing/ios/IosUnitTests` to import and execute the XCTests in the `.dylib`.
+
+See [[Setting-up-the-Engine-development-environment#editor-autocomplete-support]]
+for tips on setting up C/C++/Objective-C code completion and syntax highlighting
+when working on the engine and tests.
+
+XCTests are executed during pre-submit on our CI system when submitting PRs to
+the `flutter/engine` repository.
+
+#### XCTest
+
+[XCTest](https://developer.apple.com/documentation/xctest) is the standard way
+of creating unit tests in Xcode projects. Since iOS has x86 simulators and
+since we can build x86 engines, we can execute the XCTests directly on macOS
+in a headless simulator using the real iOS SDK.
```

Comment:
~~Since iOS has x86 simulators and since we can build x86 engines, we can execute the XCTests directly on macOS in a headless simulator using the real iOS SDK.~~

Since iOS has x86 simulators we execute the XCTests directory on macOS with the simulator using an x86 engine.

(It's not headless, you can watch it in the background)

Reply:
I think it's headless by default no? i.e. if you don't open the simulator yourself, there won't be any simulator windows that appear when you're running the test.

Reply:
There is no way to get it to run headless that I know of.  It will always launch the simulator app, although it won't steal focus if it's already open below.

Reply:
![Kapture 2020-06-01 at 16 06 55](https://user-images.githubusercontent.com/156888/83463137-fdf83680-a421-11ea-8740-725e5ad90ef5.gif)

I'm seeing this when I run (no simulator in dock). Is that what you're seeing too?